src/libraries/System.Runtime.InteropServices/src/System/Runtime/InteropServices/PosixSignalRegistration.Windows.cs

```diff
@@ -1,18 +1,134 @@
+            {
+                throw new ArgumentNullException(nameof(handler));
+            }
+
+            lock (SyncObj)
+            {
+                switch (signal)
+                {
+                    case PosixSignal.SIGINT:
+                    case PosixSignal.SIGQUIT:
+                    case PosixSignal.SIGTERM:
+                    case PosixSignal.SIGHUP:
+                        break;
+
+                    default:
+                        throw new PlatformNotSupportedException();
+                }
+
+                if (s_handlers.Count == 0 &&
+                    !Interop.Kernel32.SetConsoleCtrlHandler(&HandlerRoutine, Add: true))
+                {
+                    throw Win32Marshal.GetExceptionForLastWin32Error();
+                }
+
+                var token = new Token(signal, handler);
+                s_handlers.Add(token);
+                return new PosixSignalRegistration(token);
+            }
+        }
+
+        public partial void Dispose()
+        {
+            lock (SyncObj)
+            {
+                if (_token is Token token)
+                {
+                    _token = null;
+
+                    s_handlers.Remove(token);
+                    if (s_handlers.Count == 0 &&
+                        !Interop.Kernel32.SetConsoleCtrlHandler(&HandlerRoutine, Add: false))
+                    {
+                        throw Win32Marshal.GetExceptionForLastWin32Error();
+                    }
+                }
+            }
+        }
+
+        [UnmanagedCallersOnly]
+        private static Interop.BOOL HandlerRoutine(int dwCtrlType)
+        {
+            PosixSignal signal;
+            switch (dwCtrlType)
+            {
+                case Interop.Kernel32.CTRL_C_EVENT:
+                    signal = PosixSignal.SIGINT;
+                    break;
+
+                case Interop.Kernel32.CTRL_BREAK_EVENT:
+                    signal = PosixSignal.SIGQUIT;
+                    break;
+
+                case Interop.Kernel32.CTRL_SHUTDOWN_EVENT:
+                    signal = PosixSignal.SIGTERM;
+                    break;
+
+                case Interop.Kernel32.CTRL_CLOSE_EVENT:
+                    signal = PosixSignal.SIGHUP;
+                    break;
+
+                default:
+                    return Interop.BOOL.FALSE;
+            }
+
+            List<Token>? tokens = null;
+            lock (SyncObj)
+            {
+                foreach (Token token in s_handlers)
+                {
+                    if (token.Signal == signal)
+                    {
+                        (tokens ??= new()).Add(token);
+                    }
+                }
+            }
+
+            if (tokens is null)
+            {
+                return Interop.BOOL.FALSE;
+            }
+
+            var context = new PosixSignalContext(signal);
+            foreach (Token handler in tokens)
+            {
+                handler.Handler(context);
```

Comment:
The Unix implementation uses synchronization to ensure the handlers don't get called after `Dispose` returns.
This way it is guaranteed when the handler can no longer cancel the signal.

Do we want to add it here too? Or maybe we want to remove it from the Unix implementation?

Reply:
That's a reasonable goal, and it's something that's been important for CancellationTokenRegistration.Dispose.  However, my expectation is that the predominant use case here is different, with someone registering for a signal and then either never unregistering or doing so at the end of the program, so I'm not sure how important that guarantee is.  There's also a potential for deadlock that can result, which is why CTR subsequently added an Unregister that didn't have the synchronization. 

I'll merge this and we can subsequently decide whether to add it to the Windows implementation or remove it from the Unix one.  I agree we should keep it the same across them.

Reply:
@tmds, thinking about it more, I think we should remove the synchronization from the Unix implementation rather than adding it to the Windows one.  I expect it'll be the minority case that someone needs to remove such a handler, and even more rare that the code following that removal depends on the handler not running, and in the case where it does, the code itself can synchronize if really necessary.

On top of that, it's not hard to come up with cases where this could deadlock, e.g. one thread takes a lock and does Console.CancelKeyPress -= handler concurrently with their handler running and trying to take that same lock... locking as the Unix implementation does now could cause that to deadlock.

Reply:
Actually, don't worry about it, I'll take care of it.